latest/index.bs

```diff
@@ -177,308 +177,99 @@ For this example we assume an image with 5 dimensions and axes called `t,c,z,y,x
 
 Tables {#table-layout}
 ----------------------
-The following describes the expected layout for tabular data.
-OME-NGFF tables are compatible with the [AnnData model](https://github.com/scverse/anndata).
+A table is a 2-dimensional data structure consisting of rows and columns. 
+Tables are an intuitive way of storing arbitrary data or metadata that are 
+highly structured, potentially large, and are in some way associated with an image.
+
+In OME-NGFF, a table is a Zarr group containing zero, one, or more Zarr arrays, where each
+array represents one column of the table. Columns are ordered, and each column in a 
+table MUST have the same number of rows. While the table itself MUST be 2-dimensional, 
```

Comment:
Since columns can be n-dimensional I would replace this by saying that each column should have the same shape, so to avoid the case of one column being 1d and one 2d.

Reply:
Why should this case be prohibited? As Michael pointed out, this is permitted in Matlab.

Reply:
This could make the in-memory representation of the data challenging (for instance `AnnData` or `Pandas` may not work out of the box). For instance, should a 2D column be represented as a rows that are `np.array()` or as a list of columns? In the first case writing a `Pandas` dataframe back to disk would give a list of 1d columns instead that a single 2d column.

Reply:
I don't really understand this objection. I understand that this structure cannot be stored in a Pandas dataframe, which is Python-only data structure. But AnnData has means to store 1D annotations (as Pandas dataframe, as far as I understand) as well as nD annotations (dictionary of type `string -> [np.array|<other AnnData in-memory representations>]`?) in-memory. Could you please elaborate on the problems that you see with AnnData?

Reply:
With the previous `AnnData` table proposal the main dataframe was saved as a matrix (so each column had always dimension 1), and eventual ndim annotatoins were saved in `obsm`, that was not the main matrix but present in a different zarr group (the `obsm` group).

With the new proposal from what I understand there is just one main dataframe. So it's true that by default we could store any column that has dimension > 1 into a `obsm` in `AnnData`, but this type of conversion is not bidirectional and can induce fragmentation. For instance if the use has an `AnnData` object with an `obsm` column with dimension 1, when saving it to disk should this be saved in the separate `obsm` group or should it be saved in the main matrix? In that case upon re-reading there is no way to know that before it was in `obsm`.

Reply:
You're right, I see your point. From the [AnnData documentation](https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html#Observation/variable-level-matrices) it seemed to me as if only multidimensional annotations can be stored in `obsm/varm`, however AnnData lets me store a 1d array into the `obsm` dictionary. I'm a bit puzzled by this: why would I want to store a 1d-array in `obsm` instead of `obs`?

Anyway, it is still possible to have `obs/obsm/obsp` as three different tables. This is entirely up to the AnnData-spec on top of this more minimal table spec, my AnnData implementation is merely a suggestion.

Reply:
I also see Luca's point. If it's hard for Python, it's likely hard for other languages, too. Allowing multidimensional columns may just be a quirky Matlab thing, since, as far as I know, R doesn't allow it. The benefits of the "nice-to-have" 2D columns need to be weighed against the costs of messier code to support this feature. (And, as Luca pointed out, the incompatibility with AnnData.)

I would change the second paragraph to this:
"In OME-NGFF, a table is a Zarr group containing zero, one, or more Zarr arrays, where each array represents one column of the table. Columns are ordered, and MUST be 1-dimensional. Each column in a table MUST have the same number of rows. A row of the table is the set of arrays obtained by slicing all columns at the same index in the 0-th dimension."

Reply:
Exactly, I would store them as separate tables.

Reply:
So do we agree to restrict the spec to 1D columns?

Reply:
I think that restricting to 1D columns would severely break AnnData compatibility. Rather, I support Luca's suggestion of having a table (in our definition) for `obs`, `obsm`, and `obsp`, each, where one can store 1D, nD with n > 1, and quadratic arrays, respectively.

Reply:
I would prefer 1D columns of tuples over having 2D columns. For now this could be restricted to homogeneous tuples. This would allow these to be exactly mappable to a higher dimensional arrays. However, in the future, heterogenous tuples, structs with arbitrary fields, may be useful to have.

For example, instead of having a 5 row by 6 column matrix, we have a vector, a one-dimensional array, of 6-tuples.

```python
In [26]: import numpy as np

In [27]: A = np.reshape(np.arange(30), (5,6))

In [28]: A # 2D matrix representation
Out[28]:
array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23],
       [24, 25, 26, 27, 28, 29]])

In [29]: B = [tuple(row) for row in A]

In [30]: B # 6-tuple representation
Out[30]:
[(0, 1, 2, 3, 4, 5),
 (6, 7, 8, 9, 10, 11),
 (12, 13, 14, 15, 16, 17),
 (18, 19, 20, 21, 22, 23),
 (24, 25, 26, 27, 28, 29)]

In [31]: np.array(B) # Easy conversion from 6-tuple representation back to a 2D matrix
Out[31]:
array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23],
       [24, 25, 26, 27, 28, 29]])
```